libnetwork/drivers/bridge/setup_ip_tables_linux_test.go

```diff
@@ -184,3 +203,148 @@ func TestSetupIP6TablesWithHostIPv4(t *testing.T) {
+			enableIPMasquerade: true,
+			wantIPv4Masq:       true,
+		},
+		{
+			desc:               "IPv4 SNAT, IPv6 disabled",
+			enableIPTables:     true,
+			enableIPMasquerade: true,
+			hostIPv4:           hostIPv4,
+			wantIPv4Snat:       true,
+		},
+		{
+			desc:               "IPv4 masquerade, IPv6 masquerade",
+			enableIPTables:     true,
+			enableIP6Tables:    true,
+			enableIPv6:         true,
+			enableIPMasquerade: true,
+			wantIPv4Masq:       true,
+			wantIPv6Masq:       true,
+		},
+		{
+			desc:               "IPv4 SNAT, IPv6 masquerade",
+			enableIPTables:     true,
+			enableIP6Tables:    true,
+			enableIPv6:         true,
+			enableIPMasquerade: true,
+			hostIPv4:           hostIPv4,
+			wantIPv4Snat:       true,
+			wantIPv6Masq:       true,
+		},
+	} {
+		t.Run(tc.desc, func(t *testing.T) {
+			defer netnsutils.SetupTestOSContext(t)()
+			dc := &configuration{
+				EnableIPTables:  tc.enableIPTables,
+				EnableIP6Tables: tc.enableIP6Tables,
+			}
+			r := &testRegisterer{t: t}
+			if err := Register(r, map[string]interface{}{netlabel.GenericData: dc}); err != nil {
+				t.Fatal(err)
+			}
+			if r.d == nil {
+				t.Fatal("testRegisterer.RegisterDriver never called")
+			}
+			nc := &networkConfiguration{
+				BridgeName:         br,
+				AddressIPv4:        brIPv4,
+				AddressIPv6:        brIPv6,
+				EnableIPv6:         tc.enableIPv6,
+				EnableIPMasquerade: tc.enableIPMasquerade,
+				HostIPv4:           tc.hostIPv4,
+			}
+			ipv4Data := []driverapi.IPAMData{{Pool: maskedBrIPv4, Gateway: brIPv4}}
+			ipv6Data := []driverapi.IPAMData{{Pool: maskedBrIPv6, Gateway: brIPv6}}
+			if !nc.EnableIPv6 {
+				nc.AddressIPv6 = nil
+				ipv6Data = nil
+			}
+			if err := r.d.CreateNetwork("nattest", map[string]interface{}{netlabel.GenericData: nc}, nil, ipv4Data, ipv6Data); err != nil {
+				t.Fatal(err)
+			}
+			defer func() {
+				if err := r.d.DeleteNetwork("nattest"); err != nil {
+					t.Fatal(err)
+				}
+			}()
+			// Log the contents of all chains to aid troubleshooting.
+			for _, ipv := range []iptables.IPVersion{iptables.IPv4, iptables.IPv6} {
+				ipt := iptables.GetIptable(ipv)
+				for _, table := range []iptables.Table{iptables.Nat, iptables.Filter, iptables.Mangle} {
+					out, err := ipt.Raw("-t", string(table), "-S")
+					if err != nil {
+						t.Error(err)
+					}
+					t.Logf("%s: %s %s table rules:\n%s", tc.desc, ipv, table, string(out))
+				}
+			}
+			for _, rc := range []struct {
+				want bool
+				rule iptRule
+			}{
+				{tc.wantIPv4Masq, iptRule{iptables.IPv4, iptables.Nat, "POSTROUTING", []string{"-s", maskedBrIPv4.String(), "!", "-o", br, "-j", "MASQUERADE"}}},
+				{tc.wantIPv4Snat, iptRule{iptables.IPv4, iptables.Nat, "POSTROUTING", []string{"-s", maskedBrIPv4.String(), "!", "-o", br, "-j", "SNAT", "--to-source", hostIPv4.String()}}},
+				{tc.wantIPv6Masq, iptRule{iptables.IPv6, iptables.Nat, "POSTROUTING", []string{"-s", maskedBrIPv6.String(), "!", "-o", br, "-j", "MASQUERADE"}}},
```

Comment:
Is the order of these rules significant? The tests in [libnetwork/firewall_linux_test.go](https://github.com/moby/moby/blob/master/libnetwork/firewall_linux_test.go) assert that rules are inserted in the desired order; you may find some inspiration there.

FYI:
- Feel free to import `gotest.tools/v3/assert`. The only reason it is not already used in this file is historical: libnetwork used to be a separate project (https://github.com/moby/libnetwork) with different opinions on testing.
- `netnsutils.SetupTestOSContext(t)` moves the calling goroutine into a fresh, unshared network namespace which guarantees a blank-slate state for the iptables rulesets.

Reply:
> Is the order of these rules significant?

No.  I added a comment to hopefully make this clear to future readers.

>     * Feel free to import `gotest.tools/v3/assert`.

Done, thanks for the pointer.

>     * `netnsutils.SetupTestOSContext(t)` moves the calling goroutine into a fresh, unshared network namespace which guarantees a blank-slate state for the iptables rulesets.

I added a comment explaining why that isn't called here.